### PR TITLE
Database assetstore

### PR DESCRIPTION
**Pull request by manthey**
The database assetstore supports Mongo and sqlalchemy SQL types (provided dependencies are installed).  For postgres, all non-volatile functions are allowed (as for db items).  Each file in the assetstore references a specific table or collection.  The download endpoint uses extraParameters to perform full selects on the database (with the same basic parameters as used by the item select endpoint).  Default values for sort, fields, filters, limit, and format can be specified, and these can be overridden in the query.

To support range-requests, we preload the response to calculate its length.  This is inefficient and a waste of memory.

The mongo database connector now caches field information.  Since it inspects the entire collection to determine available fields, this is an expensive operation.  This could still use improvement.

If _id is not explicitly included in a field list for Mongo, it is excluded, so as to only return the fields that were actually specified.

Tests are needed.

### CODE REVIEW

**Comment by manthey** [comment]
Our files do not have a known size (since it depends on the data in the database).  I've stored a 0 for the size value, but this has some weird effects: the UI shows a size of 0, where it really shouldn't show anything.  Range requests are ignored, as the cherrypy utility for calculating ranges is called with file['size'].  Ideally, the file['size'] would have a special value (`None`?) indicating that it is a dynamic, unknown size, but some changes to Girder core will be required for this.

**Comment by manthey** [comment]
Currently, the database assetstore references a specific database (as well as the connection information for the database).  One database assetstore can handle any number of tables, but you need a separate assetstore for each database.  Instead, if the assetstore is just the database connector, then the files can reference both the table AND the database, and one assetstore can reference all databases and tables on that connector.  

This will change the import function to list all databases and all tables in all databases (which will take longer to query).  Databases will be added as folders, with the tables as currently done.  This will allow top-level containers (collections and users) to be the appropriate parent for import as well.

**Comment by aashish24** [comment]
@manthey looks like a great addition. Let's touch base on it some more tomorrow.

**Comment by ronichoudhury** [comment]
@manthey, did you and @aashish24 get a chance to talk about this? Did any new ideas or potential pitfalls with the current approach come up?

**Comment by manthey** [comment]
We did.  Basically, I'll move the special item endpoints to be special file
endpoints that will allow querying and changing the individual parameters
stored in the files.

If the url for an assetstore includes the database, the behavior will be
the same as currently on this branch.  If you exclude the database in the
assetstore url, then the import will list all databases and tables.  If you
pick a collection or user to add the tables to, it will create a folder
with the name of the database.

On Mon, Jul 11, 2016 at 10:56 AM, Roni Choudhury <notifications@github.com>
wrote:

> @manthey <https://github.com/manthey>, did you and @aashish24
> <https://github.com/aashish24> get a chance to talk about this? Did any
> new ideas or potential pitfalls with the current approach come up?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenGeoscience/girder_db_items/pull/8#issuecomment-231759625>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AIX_R3Cd8CTH9LTxJRYRNnNuPmjUB-T8ks5qUlmigaJpZM4JEomK>
> .
>



-- 
David Manthey
R&D Engineer
Kitware Inc.
(518) 881-4439

**Comment by ronichoudhury** [comment]
Sounds great!

I forgot to mention, but @jeffbaumes had mentioned another reason multiple assetstores for a single logical database would be necessary: if the different tables had different permissions on them, you would need separate assetstores, each instantiated with different credentials.

Thanks for this work, it's very cool :sunglasses:

**Comment by manthey** [comment]
Left to do for this branch:
- [ ] Allow assetstores to not specify the database (and the import process to do so)
- [ ] Hide usernames and passwords in the display of the assetstore information
- [ ] Add more server tests (client tests will be a separate PR)

**Comment by alex-r-bigelow** [comment]
I realize this is still WIP, but in case it's useful, I'm getting this error from grunt (just tried it on a fresh girder installation):

```
Running "plugin:database_assetstore" (plugin) task
Verifying properties pluginDir, plugin.database_assetstore.tasks exist in config...ERROR
>> Unable to process task.
Warning: Required config property "plugin.database_assetstore.tasks" missing. Use --force to continue.
```

**Comment by manthey** [comment]
I've renamed the plugin to `database_assetstore`.  This means that you have to name the plugin directory in Girder to `database_assetstore` rather than `girder_db_items`, or it throws this error.  The plugin's Gruntfile.js could be changed to have both names in it, or I could make the grunt file figure out its local directory.

**Comment by manthey** [comment]
@ronichoudhury @aashish24 This PR is ready for review.  There are server tests for all of the code, except for some of the MongoDB connector and one edge case in the sqlalchemy connector (98% coverage of server code).  I'll add client tests next (in a separate PR, unless you feel it is necessary to approve this one or the review takes long enough).

**Comment by manthey** [comment]
Now working in Python 2.7 and 3.4.

**Comment by alex-r-bigelow** [comment]
It looks like the `"plugin.database_assetstore.tasks" is missing` problem has been fixed at this point. Thanks!

**Comment by aashish24** [comment]
@manthey let's get it merged, I have some minor comments but we could talk about it when we meet next week.

**Comment by zachmullen** [comment]
Sorry for the delay, I'm QAing this branch now.

**Comment by aashish24** [comment]
that's great @zachmullen thanks.

File: .travis.yml [hunk 1]
```diff
@@ -64,12 +73,15 @@ before_script:
...
+    - gunzip -c data/plugins/database_assetstore/testdb.sql.gz | psql -U postgres sampledb
     - psql -U postgres -h 127.0.0.1 sampledb -c "SELECT table_type,table_name FROM information_schema.tables WHERE table_schema='public' order by table_name;" | cat
+    # Load mongo and show that we loaded the database properly
+    - bunzip2 -c data/plugins/database_assetstore/mongodb.permits.json.bz2 | mongoimport --db boston --collection permits
+    - mongo boston --eval 'db.permits.count()'
+
     # Only some functions are available in the version that Travis has.
     # - psql -U postgres -h 127.0.0.1 sampledb -c "SELECT lower(proname), provolatile FROM pg_proc order by lower(proname);" | cat
-    - JASMINE_TIMEOUT=15000 ctest -VV -R '(coverage_reset|py_coverage_combine|girder_db_items|py_coverage$|py_coverage_xml)'
+    - JASMINE_TIMEOUT=15000 ctest -VV -R '(coverage_reset|coverage_combine|database_assetstore|coverage$|coverage_xml)'
```
Comment (zachmullen) [inline .travis.yml]: We recently added some handy config vars to girder to avoid these sorts of regexes. You can add the following config-time definitions:

    -DRUN_CORE_TESTS:BOOL=OFF -DBUILD_JAVASCRIPT_TESTS:BOOL=OFF -DTEST_PLUGINS:STRING=database_assetstore

File: plugin.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
-    "name": "Database Items",
-    "description": "Provide database access via items.",
+    "name": "Database Assetstore",
```
Comment (zachmullen) [inline plugin.json]: These names are conventionally sentence case.

File: plugin_tests/assetstore_test.py [hunk 3]
```diff
@@ -0,0 +1,633 @@
...
+        self.dbParams3 = {
+            'type': AssetstoreType.DATABASE,
+            'name': 'One Mongo Assetstores',
+            'dburi': os.environ.get('GIRDER_TEST_DB')
+        }
+
+    def testAssetstoreCreate(self):
+        # Make sure admin access required
+        resp = self.request(path='/assetstore', method='POST',
+                            params=self.dbParams)
```
Comment (zachmullen) [inline plugin_tests/assetstore_test.py]: A better assertion would be to request this as `self.user` and then assert 403 status.

File: plugin.json (hunk 2 above)
Comment (manthey) [inline plugin.json]: Okay.  If we care, a few other plugins are also title cased (Item Previews and Google Analytics (if it counts, since that could be viewed as a proper name)).

Comment (zachmullen) [inline plugin.json]: We should change Item Previews, I'd say we should keep Google Analytics capitalized since it's proper.

File: plugin_tests/assetstore_test.py [hunk 4]
```diff
@@ -0,0 +1,633 @@
...
+            'format': 'list',
+            'fields': 'town,pop2010',
+            'limit': '10'
+        }
+        resp = self.request(
+            path='/database_assetstore/%s/import' % str(assetstore1['_id']),
+            method='PUT', user=self.admin, params=params)
+        self.assertStatusOk(resp)
+        townItem = list(self.model('item').textSearch('towns', user=self.admin,
+                                                      limit=1))[0]
```
Comment (zachmullen) [inline plugin_tests/assetstore_test.py]: I think most girder devs have expressed a preference against visual indentation of params due to readability concerns.

File: plugin_tests/dbs_mongo_test.py [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+        self.assertStatusOk(resp)
+        self.assertNotEqual(resp.json['data'][0][2], '1-7FAM')
+        params['filters'] = json.dumps([{
+            'field': 'occupancytype', 'operator': 'not_is', 'value': None}])
+        resp = self.request(path='/file/%s/database/select' % (
+            self.dbFileId, ), user=self.admin, params=params)
+        self.assertStatusOk(resp)
+        self.assertNotEqual(resp.json['data'][0][2], None)
+
+    def XtestMongoDatabaseSelectClient(self):
```
Comment (zachmullen) [inline plugin_tests/dbs_mongo_test.py]: Did you mean to remove this from the test suite?

File: plugin_tests/dbs_mongo_test.py [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+            'params': params
+        })
+        slow.start()
+        # Wait for the query to start
+        while not sessions['test']['used'] and slow.is_alive():
+            time.sleep(0.05)
+        # Sending a normal request should cancel the slow one and respond
+        # promptly
+        resp = self.request(path='/file/%s/database/select' % (
+            self.dbFileId, ), user=self.admin, params=params)
```
Comment (zachmullen) [inline plugin_tests/dbs_mongo_test.py]: If you are only passing one argument to string formatting, you don't need a tuple.

File: plugin_tests/dbs_mongo_test.py [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+            'town',
+            {'func': 'pg_sleep', 'param': [40]},
+        ])
+        slowParams['limit'] = 500
+        slowResults = {}
+
+        def slowQuery(params):
+            try:
+                self.request(path='/file/%s/database/select' % (
+                    self.dbFileId, ), user=self.admin, params=slowParams)
```
Comment (zachmullen) [inline plugin_tests/dbs_mongo_test.py]: If you are only passing one argument to string formatting, you don't need a tuple.

File: plugin_tests/dbs_mongo_test.py [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+            'field': 'occupancytype', 'operator': 'not_is', 'value': None}])
+        resp = self.request(path='/file/%s/database/select' % (
+            self.dbFileId, ), user=self.admin, params=params)
+        self.assertStatusOk(resp)
+        self.assertNotEqual(resp.json['data'][0][2], None)
+
+    def XtestMongoDatabaseSelectClient(self):
+        params = {'sort': 'town', 'limit': 1, 'clientid': 'test'}
+        resp = self.request(path='/file/%s/database/select' % (
+            self.dbFileId, ), user=self.admin, params=params)
```
Comment (zachmullen) [inline plugin_tests/dbs_mongo_test.py]: string formatting tuple

File: plugin_tests/dbs_mongo_test.py [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+        self.assertStatusOk(resp)
+        from girder.plugins.database_assetstore import dbs
+        sessions = dbs.base._connectorCache[self.dbFileId].sessions
+        # We should be tracking the a session for 'test'
+        self.assertIn('test', sessions)
+        self.assertFalse(sessions['test']['used'])
+        last = sessions['test'].copy()
+        # A new request should update the last used time
+        resp = self.request(path='/file/%s/database/select' % (
+            self.dbFileId, ), user=self.admin, params=params)
```
Comment (zachmullen) [inline plugin_tests/dbs_mongo_test.py]: string formatting tuple

File: plugin_tests/dbs_mongo_test.py [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+        self.assertStatusOk(resp)
+        self.assertGreater(sessions['test']['last'], last['last'])
+        self.assertEqual(sessions['test']['session'], last['session'])
+        # Artifically age the last session and test that we get a new session
+        last = sessions['test'].copy()
+        sessions['test']['last'] -= 305  # 300 is the default expiry age
+        resp = self.request(path='/file/%s/database/select' % (
+            self.dbFileId, ), user=self.admin, params=params)
```
Comment (zachmullen) [inline plugin_tests/dbs_mongo_test.py]: string formatting tuple

File: plugin_tests/file_test.py [hunk 11]
```diff
@@ -101,370 +136,349 @@ def setUp(self):
...
-        resp = self.request(method='PUT', path='/item/%s/database/refresh' % (
-            itemId, ), user=self.admin)
+            resp = self.request(path='/file/%s/database/fields' % (
+                fileId, ), user=self.admin)
+
+    def testFileDatabaseRefresh(self):
+        fileId, fileId2, fileId3 = self._setupDbFiles()
+
+        resp = self.request(method='PUT', path='/file/%s/database/refresh' % (
+            fileId3, ), user=self.admin)
```
Comment (zachmullen) [inline plugin_tests/file_test.py]: here too

File: plugin_tests/dbs_mongo_test.py (hunk 5 above)
Comment (manthey) [inline plugin_tests/dbs_mongo_test.py]: More that it was never added...

We don't currently have support for the concept of a clients in the Mongo connector.  In Postgres, this can be used to cancel a query when it is replaced by a fresher query (for when the user no longer cares about the answer to the first query).

Comment (zachmullen) [inline plugin_tests/dbs_mongo_test.py]: OK, just making sure it was intentional.

Comment (manthey) [inline plugin_tests/dbs_mongo_test.py]: I'll remove it.

File: server/assetstore.py [hunk 12]
```diff
@@ -0,0 +1,434 @@
...
+from girder.models.model_base import GirderException, ValidationException
+from girder.utility.abstract_assetstore_adapter import AbstractAssetstoreAdapter
+from girder.utility.model_importer import ModelImporter
+
+from .dbs import getDBConnectorClassFromDialect, databaseFromUri, \
+    getDBConnectorClass
+from .query import dbFormatList, queryDatabase, preferredFormat
+
+
+dbInfoKey = 'databaseMetadata'
```
Comment (zachmullen) [inline server/assetstore.py]: If this is just a constant, can we make it `ALL_CAPS`?

File: plugin_tests/dbs_mongo_test.py (hunk 7 above)
Comment (manthey) [inline plugin_tests/dbs_mongo_test.py]: True, but apparently I use them a lot in the tests, mostly to facilitate line breaks (this was written in the 80-column limit days).  Is it worth changing them all?

File: server/assetstore.py [hunk 13]
```diff
@@ -0,0 +1,434 @@
...
+            cherrypy.response.headers['Content-Disposition'] = \
+                'attachment; filename="%s"' % file['name']
+        cherrypy.response.headers['Accept-Ranges'] = 'bytes'
+        if (endByte is not None and offset is not None and
+                file['size'] is not None):
+            cherrypy.response.headers['Content-Length'] = max(
+                endByte - offset, 0)
+            if (offset or endByte < file['size']) and file['size']:
+                cherrypy.response.headers['Content-Range'] = \
+                    'bytes %d-%d/%d' % (offset, endByte - 1, file['size'])
```
Comment (zachmullen) [inline server/assetstore.py]: Should we be passing `Accept-Ranges`? Do byte offsets actually work?

File: server/assetstore.py [hunk 14]
```diff
@@ -0,0 +1,434 @@
...
+            # record from the file.
+            #   This intentionally encodes extraParameters as JSON.  It could
+            # pass it as a python dictionary or encode it as a url query string,
+            # but another Girder plugin has expressed a preference for JSON as
+            # the de facto standard for extraParameters.
+            downloadFunc = self.downloadFile(
+                file.copy(), headers=False, extraParameters=json.dumps({
+                    'limit': 1}))
+            # Test the download without keeping it
+            [None for chunk in downloadFunc()]
```
Comment (zachmullen) [inline server/assetstore.py]: I think some PEP or another dictates that list comprehension syntax shouldn't be used for side-effects.

File: server/assetstore.py (hunk 13 above)
Comment (manthey) [inline server/assetstore.py]: Yes, byte offsets work.

File: server/assetstore.py (hunk 12 above)
Comment (manthey) [inline server/assetstore.py]: Sure.

File: plugin_tests/dbs_mongo_test.py (hunk 7 above)
Comment (zachmullen) [inline plugin_tests/dbs_mongo_test.py]: If it's more trouble than it's worth, it's no big deal to leave them in.  But the same could be achieved by just using parens and omitting the comma that makes it a tuple.

File: server/dbs/__init__.py [hunk 15]
```diff
@@ -17,9 +17,10 @@
...
-import sqlalchemydb  # noqa
-import postgres_sqlalchemy  # noqa
-import mongo  # noqa
+from .base import (   # noqa
+    getDBConnectorClass, getDBConnector, getDBConnectorClassFromDialect,
+    clearDBConnectorCache, FilterOperators, DatabaseConnectorException,
+    databaseFromUri)
+from . import sqlalchemydb  # noqa
+from . import postgres_sqlalchemy  # noqa
+from . import mongo  # noqa
```
Comment (zachmullen) [inline server/dbs/__init__.py]: I'm guessing flake8 complains about these as unused imports, but this is totally valid behavior that is very common. Is there something easy we can do besides silencing flake8 to make it clear these are being imported just to be in the module namespace?

File: server/assetstore.py (hunk 13 above)
Comment (manthey) [inline server/assetstore.py]: They are tested here: https://github.com/OpenGeoscience/girder_db_items/blob/db-as-assetstore/plugin_tests/assetstore_test.py#L392

File: server/assetstore.py (hunk 14 above)
Comment (manthey) [inline server/assetstore.py]: Would you prefer
```
for chunk in downloadFunc():
  pass
```

File: server/dbs/__init__.py (hunk 15 above)
Comment (manthey) [inline server/dbs/__init__.py]: Adding an `__all__ =` line would prevent the local unused errors.  I think that will take care of it.

File: server/dbs/postgres_sqlalchemy.py [hunk 16]
```diff
@@ -84,6 +83,10 @@ def _isFunctionAllowed(self, funcname):
                 if (func[1] not in ('i', 's') and
                         self._allowedFunctions.get(func[0], False)):
                     self._allowedFunctions[func[0]] = False
+            # Whitelist some functions if they exist
+            for func in funcs:
+                if func[0] in ('pg_sleep', ):
+                    self._allowedFunctions[func[0]] = True
```
Comment (zachmullen) [inline server/dbs/postgres_sqlalchemy.py]: What's the significance of `pg_sleep`?

Comment (manthey) [inline server/dbs/postgres_sqlalchemy.py]: I want to be able to cancel slow queries.  This has to do with the use case of the user selecting one thing and then selecting another, so the first query could be cancelled as it is no longer needed.  It is difficult to make a query to a immutable function that is both slow enough to ensure we have time to cancel it and is consistent in allowing cancellation (a lot of the PostGIS functions call some C code which doesn't get cancelled during certain phases of the query).  `pg_sleep` is a simple function that is slow and cancellable.  It is not marked as an immutable function in postgres since it has a server side effect, but it is non-volatile.  By white-listing it, I can use it in the cancel-query test.

**Comment by manthey** [comment]
I've pushed a commit that should address your comments up to this point.

File: server/rest.py [hunk 17]
```diff
@@ -22,351 +22,127 @@
...
 
-class DatabaseItemResource(Item):
+class DatabaseFileResource(File):
 
     def __init__(self, apiRoot):
-        # Don't call the parent (Item) constructor, to avoid redefining routes,
+        # Don't call the parent (File) constructor, to avoid redefining routes,
         # but do call the grandparent (Resource) constructor
-        super(Item, self).__init__()
+        super(File, self).__init__()
```
Comment (zachmullen) [inline server/rest.py]: Maybe instead of having this descend from `File`, we just pass `info['apiRoot'].file` into it and have it tack stuff on? Then we don't have to deal with the constructor messiness.

```
class DatabaseFileResource(ModelImporter):
    def __init__(self, fileResource):
        fileResource.route(...)
```

File: server/dbs/postgres_sqlalchemy.py (hunk 16 above)
Comment (zachmullen) [inline server/dbs/postgres_sqlalchemy.py]: Won't this make it so users are able to call this even in production? Is that something we are OK with allowing? My only worry is it could be used to just leave connections open to the database for indefinite periods.

File: server/rest.py (hunk 17 above)
Comment (manthey) [inline server/rest.py]: I expect to have access to Resource class methods (`getBodyJson`, `getCurrentUser`, etc.).  I'd either have to change all of those to use `self.fileResource.<function>` or do something like `self.getBodyJson = self.fileResource.getBodyJson` for each function in the `__init__` method.

File: server/dbs/postgres_sqlalchemy.py (hunk 16 above)
Comment (manthey) [inline server/dbs/postgres_sqlalchemy.py]: Good point.  I'll change things so it is only whitelisted in the tests, not in general.

File: server/rest.py (hunk 17 above)
Comment (zachmullen) [inline server/rest.py]: Ah, I see. `getBodyJson`, `getCurrentUser` et al also exist in unbound forms in the `girder.api.rest` module, but I'm fine either way.

Comment (manthey) [inline server/rest.py]: Another technique would be to remove the `DatabaseFileResource` class, make the `__init__` method into an `addFileRoutes` function, and decorate each route with `@boundHandler()`.

Do you have a preference to resolve this?

Comment (zachmullen) [inline server/rest.py]: Let's go ahead and use `boundHandler`, may as well get some mileage out of that feature.

Comment (manthey) [inline server/rest.py]: Right-o.

**Comment by zachmullen** [comment]
LGTM
